### DISCUSSION TITLE
how to add code editor/runner?

### DISCUSSION THREAD

**Discussion by tsdexter** [hrgdavor/jscadui#105] [Q&A]
Hi @hrgdavor , not a question/issue specific to this app - but can you point me in the right direction where/how you add a code editor and how you run the code browser side (without risk of XSS etc, ideally)? I want to implement a code editor in a web page that also runs the code similar to what you've done with this app.

**Comment by hrgdavor** [comment]
@tsdexter I am glad you asked. My intent in this repo is to share experiences code or components on how to make interesting stuff for jscad and beyond.

I am no expert at this, but I did play with it for some time. There are two ways to execute code in isolation: worker (done in jscad as it does not need DOM access) or in an iframe like did in this [explainer for jsx](https://hrgdavor.github.io/jsx6/demistify/) (that is bit out of date, as it does not cover jsx-runtime and jsx-dev-runtime).

much of the code is in progress so much so that only few things got a npm release.

This package handles `require` https://github.com/hrgdavor/jscadui/tree/main/packages/require and if you want to run ESM code you must use something to transform code to cjs before eval (I use babel in jscad.app currently) and if you do taht as a bonus you can execute typescript also ... actually you can mix  cjs,esm,typescript files this way because each is converted to cjs before eval. There are some gems in the code like a trick how to execute the scripts and keep correct line numbers in stack traces (what you can find online will all have stack trace offset by few lines).... bla .. bla ..

Anyway, I will be glad to help, and in the process I will likely have more ideas what to put in readme for others to find too.

**Reply by tsdexter** [reply]
Thanks a lot! I'll start taking a look around. The place that I plan to run the user generated code is already in a web worker. So is it safe to just use `eval` in a web worker - it's not going to be prone to XSS or other potential issues for user code?

**Reply by hrgdavor** [reply]
I have not looked it too deeply, but there are some risks with web workers too, although I thought they were less risky than eval in main thread. 
I googled a bit, so I suggest the same. If you find something interesting please share here if you have time. I would like to gather such info too, but at the moment am not testing security side of web workers.

**Reply by tsdexter** [reply]
I have also read that it's **not as risky** because it doesn't have access to the DOM. However, it does still have access to fetch, so can cause some issues, but nothing the user can't do on their own in developer tools, so it's generally OK with _same-user_ generated code, but still potentially problematic with _other-user_ generated code, such as if it's stored in a database and shared amongst the userbase. 

I'll continue playing around with it and report back anything here. Right now, I'm using `js-interpreter` but it's much slower than native.

**Reply by hrgdavor** [reply]
looks like for some use cases iframe could help with security, as it can have a separate thread, and definitely a separate env (no sharing auth in fetch)
and you can communicate also via postMessage.
https://stackoverflow.com/questions/11510483/will-a-browser-give-an-iframe-a-separate-thread-for-javascript

**Reply by tsdexter** [reply]
I'll test that out when I get back to it. It might be good in my case as the user-code would generate JSON/serializable data that I could pass through with `postMessage` without issue. 

I was also looking at just "disabling" `fetch` by a simple string search of `fetch` within the user generated code and throwing an error/no running `eval` if `fetch` is included in the code.

**Reply by hrgdavor** [reply]
bastards can quickly go around such attempts of cleaning code :)

**Reply by tsdexter** [reply]
Yes, I'm thinking `base64` encoded strings and plenty of other workarounds. Perhaps it's best to push it on to the user responsibility by not running `eval` until the user accepts a popup confirming they've reviewed the code and allow it to run within their browser context.

**Comment by hrgdavor** [comment]
@tsdexter I accidentally had an idea and googled it a bit. It looks like there is a way for service worker to change/strip headers. 

one of links I stumbled upon : https://stackoverflow.com/questions/49503836/serviceworker-is-it-possible-to-add-headers-to-url-request

So my idea is since jscad worker uses ServiceWorker anway I can augment ServiceWorker to strip cookie and Authorization headers from requests, thus nerfing xss attempts.

I will definitely try to access authorised url for test from worker, and then try it also with ServiceWorker striping headers.
